### ISSUE TITLE
Will padding zeros for variable length input affect batch normalization?

### ISSUE THREAD

**Issue by weedwind** [SeanNaren/deepspeech.pytorch#312]
Hi, 

I saw in the dataloader that in order to form a input batch with variable length input, the code uses zeros to pad the short sequences. I am not sure if these zeros will affect training the batch normalization, since BN will include these when it computes mean and variance, and might make the variance very small to cause any problem with training?

Thank you very much for help.

**Comment by SeanNaren** [comment]
Yes and Yes! I'm working on a fix for this. This is currently why variable-length hasn't been merged, however once I implement the fix this should handle the batch norm layers.

**Comment by weedwind** [comment]
@SeanNaren Thank you very much for help. I came up with another solution, and implemented it. When you form a batch from variable length inputs, do not pad frame with zeros, but repeat the sentence until it reaches the longest length (you can use np.pad with mode = 'wrap'). You do not need to pad the labels. When you compute CTC loss, you can pass the original length of the frames, so that CTC will only consider those frames. However, batch normalization will use all the frames to compute mean and variance, and since the padded frames are real speech, it improves batch normalization. Also, with this method, you don't need bucket sampler.

**Comment by EgorLakomkin** [comment]
@weedwind, isn't retreating the sentence until the end would still affect mean/variance for batch norm (especially variance)?

**Comment by weedwind** [comment]
@EgorLakomkin I think repeating the sentence from the beginning is just like you saying this sentence multiple times. In my experiments. Since the batch does not contain zeros, it won't make the variance very small. In my experiments, this method is more stable than padding zeros, and you don't need a bucket sampler, which means you can put any variable length sequences into a batch.

**Comment by hoyden** [comment]
@weedwind You are so brilliant!! So dose it improve your experience result?

**Comment by weedwind** [comment]
@hoyden Yes, it improves a little bit on librispeech, compared with padding zeros. Also, you don't need a bucket sampler which groups similar length sequences together. You can put any variable length utterances into a batch, even though their lengths may differ a lot. 

You do not need to pad the label sequences. The CTC loss will mask the padded part of the inputs when it matches the input to the labels. But batch normalization does include the padded part when it computes mean and variance.

**Comment by weedwind** [comment]
@hoyden Be careful that at test time, you can not repeat an utterance, since it's not practical. You have to pad zeros if you want to do test by batch. But padding zeros does not matter at test time, since the model uses the running mean and variance of BN, which has already been computed at training time.

**Comment by heibaidaolx123** [comment]
Hi, guys, any update for this issue? I found that openmt and espnet have the same issue for batchnorm or layernorm.
Any idea?

**Comment by buriy** [comment]
I think you can add random bits of silence at both ends, and model will learn how to deal with that -- especially if you have noise and other augmentations added to the sounds. I do this and I think it's ok.

**Comment by amiasato** [comment]
Since this was the only place where I found anyone with the same problem, I've uploaded a gist so you may test whether this masked version of batch normalization works: https://gist.github.com/amiasato/902fc14afa37a7537386f7b0c5537741

**Comment by georgesterpu** [comment]
@amiasato it would be best if you submitted a pull request upstream, so the broader community could avail of your work.

**Comment by amiasato** [comment]
I guess I should have put a disclaimer in which I did not really use this implementation in this project, I only googled and found this thread of discussion. Unfortunately, I wouldn't even know where to put this code.

In any case, I found in the official Pytorch forums another thread with the same issue, which may be more appropriate than posting in a closed issue of a specific project: https://discuss.pytorch.org/t/masked-batch-normalization/57813/3

Sorry for the hassle.

**Comment by SeanNaren** [comment]
@amiasato thanks for the pointer! This looks really good and should make its way into upstream. It will unfortunately break backwards compatibility though so I may make it a flag for now till the next release, and then make it default behaviour

**Comment by albertz** [comment]
This has been closed as stale but the issue itself is not really resolved, right?

Are there any new insights? Has anyone compared a correct vs wrong implementation?
